File: client/js/spells.js
define(['spell',
        '../../shared/js/gametypes'], function(
        Spell,
        config) {
    
    var Spells = {
        frostnova: Spell.extend({
            init: function(target){
                this._super(Types.Entities.FROSTNOVA, target);
                this.spellType = "aoe";
            },
        }),
        frostbolt: Spell.extend({
            init: function(target){
                this._super(Types.Entities.FROSTBOLT, target);
                this.spellType = "single";
            },
        }),
        fireball: Spell.extend({
            init: function(target){
                this._super(Types.Entities.FIREBALL, target);
                this.spellType = "directional";
            },
        }),
        icebarrier: Spell.extend({
            init: function(target){
                this._super(Types.Entities.ICEBARRIER, target);
                this.spellType = "directional";
            },
        }),
        polymorph: Spell.extend({
            init: function(target){
                this._super(Types.Entities.POLYMORPH, target);
                this.spellType = "directional";
            },
        }),
        blink: Spell.extend({
            init: function(target){
                this._super(Types.Entities.BLINK, target);
                this.spellType = "directional";
            },
        }),
        getSpell: function(kind, target) {
            console.log(Types.getKindAsString(kind));
            return new this[Types.getKindAsString(kind)](target);
        },
    };

    return Spells;
});

File: client/dart/spells.dart
library spells;

import "character.dart";
import "spell.dart";
import "../shared/dart/gametypes.dart";

class Frostnova extends Spell {

  String spellType = "aoe";

  Frostnova(Character target): super(Entities.FROSTNOVA, target);
}

class Frostbolt extends Spell {

  String spellType = "single";

  Frostbolt(Character target): super(Entities.FROSTBOLT, target);
}

class Fireball extends Spell {

  String spellType = "aoe";

  Fireball(Character target): super(Entities.FIREBALL, target);
}

class IceBarrier extends Spell {

  String spellType = "single";

  IceBarrier(Character target): super(Entities.ICEBARRIER, target);
}

class Polymorph extends Spell {

  String spellType = "single";

  Polymorph(Character target): super(Entities.POLYMORPH, target);
}

class Blink extends Spell {

  String spellType = "directional";

  Blink(Character target): super(Entities.BLINK, target);
}
